### PR TITLE
Consolidate MapLayerSettings/BackgroundMapSettings

### PR DESCRIPTION
**Pull request by mdastous-bentley**
No longer sync BackgroundMapSettings.providerName/providerData with MapImagery.BaseMapProps.   MapImagery.BaseMapProps now accepts 'BackgroundMapProviderProps' which is essentially the same as the old 'BackgroundMapSettings.providerName/providerData'.  

Internally BackgroundMapProviderProps will be converted on-the-fly to MapLayersSettings by the MapTileTreeReference : the difference with the previous implementation is that we never go from MapLayerSettings to  'BackgroundMapProviderProps'.  The resulting MapLayerSettings is never persited: If ''BackgroundMapProviderProps' was used to define BaseMapProps, 'BackgroundMapProviderProps' should be persisted.

**Breaking changes:**
BackgroundMapSettings.providerName and BackgroundMapSettings.providerData are no longer part of BackgroundMapProps.

Even though 'DisplayStyleState,changeBaseMapProps' now accepts BackgroundMapProviderProps, any previous call should continue to work as before.

### CODE REVIEW

File: core/common/src/DisplayStyleSettings.ts [hunk 1]
```diff
@@ -518,7 +520,10 @@ export class DisplayStyleSettings {
     this._monochromeMode = MonochromeMode.Flat === this._json.monochromeMode ? MonochromeMode.Flat : MonochromeMode.Scaled;
 
     this._backgroundMap = BackgroundMapSettings.fromJSON(this._json.backgroundMap);
-    this._mapImagery = MapImagerySettings.fromJSON(this._json.mapImagery, this._json.backgroundMap);
+
+    // TODO
+    // Should we have a migration path here with the old JSON
```
Comment (pmconne) [inline core/common/src/DisplayStyleSettings.ts:525]: Absolutely. Saved views created before this change using old version of BackgroundMapProps must continue to load and display as they did previously.

File: core/common/src/MapImagerySettings.ts [hunk 2]
```diff
@@ -6,47 +6,221 @@
...
  * In order to provide backward compatibility the original [[BackgroundMapSettings]] are synchronized with the [[MapImagerySettings]] base layer as long as
  * the settings are compatible.
  * @beta
  */
-export interface MapImageryProps {
-  backgroundBase?: BaseLayerProps;
-  backgroundLayers?: MapLayerProps[];
-  overlayLayers?: MapLayerProps[];
+
+export interface BackgroundMapProviderProps {
```
Comment (pmconne) [inline core/common/src/MapImagerySettings.ts:21]: Objects of this type will never be saved into the iModel or into a saved view, correct?

File: core/common/src/MapImagerySettings.ts [hunk 3]
```diff
@@ -6,47 +6,221 @@
...
+  public clone(changedProps: BackgroundMapProviderProps): BackgroundMapProvider {
+    if (undefined === changedProps)
+      return this;
+
+    const props = {
+      name: undefined !== changedProps.providerName ? changedProps.providerName : this.providerName,
+      providerData: {
+        mapType: undefined !== changedProps.providerData?.mapType ? changedProps.providerData.mapType : this.mapType,
+        visible: undefined !== changedProps.providerData?.visible ? changedProps.providerData.visible : this.visible,
+        transparency: undefined !== changedProps.providerData?.transparency ? changedProps.providerData.transparency : this.transparency,
```
Comment (pmconne) [inline core/common/src/MapImagerySettings.ts:158]: I'm unclear on why BackgroundMapProps no longer has a provider name/type but still has a transparency - is it redundant with this one?

File: core/common/src/test/BackgroundMapSettings.test.ts [hunk 4]
```diff
@@ -51,32 +49,13 @@ describe("BackgroundMapSettings", () => {
 
       const expectedSettings = BackgroundMapSettings.fromJSON(expected);
       expect(settings.equals(expectedSettings)).to.be.true;
-
-      // Check synch through base map layer.
-      const mapLayer = MapLayerSettings.fromMapSettings(settings);
```
Comment (pmconne) [inline core/common/src/test/BackgroundMapSettings.test.ts:56]: Make certain the deleted tests are replaced with (1) versions that confirm the old persistent JSON representation still behaves as previously and (2) the equivalent new version behaves the same.

File: core/common/src/MapImagerySettings.ts [hunk 5]
```diff
@@ -6,47 +6,221 @@
...
+    return clone;
+  }
+
+  /** Construct from JSON, performing validation and applying default values for undefined fields. */
+  public static fromJSON(props: BackgroundMapProviderProps) {
+    return new BackgroundMapProvider(props);
+  }
+
+  /** Used to determine is a JSON struct matches BackgroundMapProviderProps  */
+  public static isMatchingProps(props: any): boolean {
```
Comment (pmconne) [inline core/common/src/MapImagerySettings.ts:173]: Return type should be `: props is BackgroundMapProviderProps`.

File: core/common/src/MapImagerySettings.ts [hunk 6]
```diff
@@ -6,47 +6,221 @@
...
+    return props?.hasOwnProperty("providerName") || props?.hasOwnProperty("providerData");
```
Comment (pmconne) [inline core/common/src/MapImagerySettings.ts:174]: Why insist an object that inherits a providerName or providerData property is not a BackgroundMapProviderProps?

File: core/frontend/src/DisplayStyleState.ts [hunk 7]
```diff
@@ -395,22 +397,30 @@ export abstract class DisplayStyleState extends ElementState implements DisplayS
...
+    } else if (BackgroundMapProvider.isMatchingProps(props)) {
+      // props is BackgroundMapProviderProps
+      if (this.settings.mapImagery.backgroundBase instanceof BackgroundMapProvider) {
+        const mapProvider = this.settings.mapImagery.backgroundBase;
+        this.settings.mapImagery.backgroundBase = mapProvider.clone(props as BackgroundMapProviderProps);
+      } else {
+        this.settings.mapImagery.backgroundBase = BackgroundMapProvider.fromJSON(props as BackgroundMapProviderProps);
+      }
     } else {
+      // props is MapLayerProps
```
Comment (pmconne) [inline core/frontend/src/DisplayStyleState.ts:416]: Not necessarily - `{ }` is a valid BackgroundMapProviderProps; in this context it means "don't change anything".

File: core/frontend/src/DisplayStyleState.ts [hunk 8]
```diff
@@ -422,16 +432,30 @@ export abstract class DisplayStyleState extends ElementState implements DisplayS
...
+      this.settings.mapImagery.backgroundBase = this.settings.mapImagery.backgroundBase.clone({providerData: {transparency}});
+      this._synchBackgroundMapImagery();
     } else {
       this.changeBaseMapProps({ transparency });
     }
   }
 
+  /** @beta  */
+  public changeBaseMapVisible(visible: boolean) {
+    // TODO: There is no way actually to change the visibility of the base map when its sets to a ColorDef.
```
Comment (pmconne) [inline core/frontend/src/DisplayStyleState.ts:450]: Seems like API shouldn't allow this then.

**Review by pmconne — COMMENTED** [review]
Could all the `if (instanceof X) else if (instanceof Y)` checks be simplified?

File: core/common/src/MapImagerySettings.ts (hunk 2 above)
Comment (mdastous-bentley) [reply core/common/src/MapImagerySettings.ts:21]: Yes it should be saved.

File: core/common/src/MapImagerySettings.ts (hunk 3 above)
Comment (mdastous-bentley) [reply core/common/src/MapImagerySettings.ts:158]: I agree its a bit confusing:
Transparency on BackgroundMapProps is an independent setting, it apply transparency on all components of mapImagery: baseMap, overlayLayers and BackgroundLayers.  Additionally, transparency can be set on independently foreach component (i.e MapLayerSettings.transparency, and now BackgroundMapProps.providerData.transparency).

Is there a business case for BackgroundMapProps.Transparency ?  I dont know, but its always been there and the map-layers widget expose it quite obviously.

File: core/common/src/MapImagerySettings.ts (hunk 5 above)
Comment (mdastous-bentley) [reply core/common/src/MapImagerySettings.ts:173]: nice typescript feature! Thanks.

File: core/common/src/MapImagerySettings.ts (hunk 6 above)
Comment (mdastous-bentley) [reply core/common/src/MapImagerySettings.ts:174]: good catch, I will use the 'in' operator.

File: core/frontend/src/DisplayStyleState.ts (hunk 8 above)
Comment (mdastous-bentley) [reply core/frontend/src/DisplayStyleState.ts:450]: Are you suggestion that `changeBaseMapVisible ` should check if `backgroundBase` is a ColorDef, and return an error/throw?

**Comment by mdastous-bentley** [comment]
> Could all the `if (instanceof X) else if (instanceof Y)` checks be simplified?

Do you mean remove them?  Thats  a side effect of having a union type for BaseLayerSettings.  Any suggestion?

File: core/common/src/DisplayStyleSettings.ts (hunk 1 above)
Comment (mdastous-bentley) [reply core/common/src/DisplayStyleSettings.ts:525]: Sent you an email related to this, I need advice on how this should be correctly done.  One problem is that `this._json`  is typed with the 'new' definition, and previous properties are not accessible.  Is casting to `any` and retrieve legacy settings a good way to solve this?

File: core/frontend/src/DisplayStyleState.ts (hunk 8 above)
Comment (pmconne) [reply core/frontend/src/DisplayStyleState.ts:449]: Ideally we don't provide users with APIs that can't perform their advertised function in the first place.

**Comment by pmconne** [comment]
> Do you mean remove them? Thats a side effect of having a union type for BaseLayerSettings. Any suggestion?

First question is whether the introduction of a new type (BackgroundMapProvider) to the union is necessary. Previously it was either a ColorDef or a MapLayerSettings, where the MapLayerSettings might have been created from a BackgroundMapProps.

If adding a third type to the union **is** necessary, can BackgroundMapProvider and MapLayerSettings be implemented such that you don't need to know which type it is when using its API?
For example, you have this:
```
  public get baseTransparency(): number {
    if  (this._backgroundBase instanceof ColorDef) {
      return (this._backgroundBase.getTransparency() / 255);
    } else if (this._backgroundBase instanceof BackgroundMapProvider) {
      return  (this._backgroundBase ).transparency;
    } else {
      return  (this._backgroundBase ).transparency;
    }
```
But the second `instanceof` check is unnecessary. Both BackgroundMapProvider and MapLayerSettings have a `transparency` property with identical semantics, so you can simplify:
```
  public get baseTransparency(): number {
    if (this._backgroundBase instanceof ColorDef)
      return this._backgroundBase.getTransparency() / 255;
    else
      return this._backgroundBase.transparency;
  }
```

File: core/common/src/DisplayStyleSettings.ts (hunk 1 above)
Comment (pmconne) [reply core/common/src/DisplayStyleSettings.ts:525]: We need to draw a distinction between changing the API and changing the persistence format. Every saved view in existence today stores their background map settings using BackgroundMapProps, so we are stuck supporting that forever - we can only add to the persistence format, not subtract. IMO you should mark BackgroundMapProps.providerName/Data as `@deprecated` (that term refers to an API, not to data) and `@internal` but not remove them.

**Comment by mdastous-bentley** [comment]
> public get baseTransparency()

If if make `BackgroundMapProvider` and `MapLayerSettings` looks the same, then I won't be able distinguish props inside `changeBaseMapProps`:
```
  public changeBaseMapProps(props: ColorDef | MapLayerProps | BackgroundMapProviderProps) {
    if (props instanceof ColorDef) {
      let transparency = 0;
      if (this.settings.mapImagery.backgroundBase instanceof ColorDef)
        transparency = this.settings.mapImagery.backgroundBase.getTransparency();
      this.settings.mapImagery.backgroundBase = props.withTransparency(transparency);
    } else if (BackgroundMapProvider.isMatchingProps(props)) {
      // props is BackgroundMapProviderProps
      if (this.settings.mapImagery.backgroundBase instanceof BackgroundMapProvider) {
        const mapProvider = this.settings.mapImagery.backgroundBase;
        this.settings.mapImagery.backgroundBase = mapProvider.clone(props);
      } else {
        this.settings.mapImagery.backgroundBase = BackgroundMapProvider.fromJSON(props);
      }
    } else {
      if (this.settings.mapImagery.backgroundBase instanceof MapLayerSettings)
        this.settings.mapImagery.backgroundBase = this.settings.mapImagery.backgroundBase.clone(props as MapLayerSettings);
      else {
        const backgroundLayerSettings = MapLayerSettings.fromJSON(props as MapLayerSettings);
        if (backgroundLayerSettings)
          this.settings.mapImagery.backgroundBase = backgroundLayerSettings;
      }
    }
    this._synchBackgroundMapImagery();
  }
```
Problem is that I need to call <class>.fromJSON based on 'type' of the props.  Do you have a workaround for this?

**Comment by pmconne** [comment]
> Problem is that I need to call the right fromJSON method based on 'type' of the props. Do you have a workaround for this?

One workaround is to not write a function that takes a variety of "props" types and has to distinguish between them. Instead, let the caller figure out what he wants the base layer to be.
For example, if you allow the base layer to be a `BackgroundMapProvider | MapLayerSettings | ColorDef`, make this function take one of those; and add functions like `BackgroundMapProvider.clone(changedProps: Partial<BackgroundMapProviderProps>)`.

File: core/common/src/test/BackgroundMapSettings.test.ts (hunk 4 above)
Comment (mdastous-bentley) [reply core/common/src/test/BackgroundMapSettings.test.ts:56]: For this particular test, there isnt much we cant test right?  Should I keep testing old JSON (i.e `{ providerName: "BingProvider" }`) and make sure it has no effect on BackgroundMapSettings anyomore?

**Pull request closed without merging**